pkg/transport/process/utils/oci_artifact_serialization.go

```diff
@@ -0,0 +1,278 @@
+// SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors.
+//
+// SPDX-License-Identifier: Apache-2.0
+package utils
+
+import (
+	"archive/tar"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"path"
+	"strings"
+
+	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/specs-go"
+	ocispecv1 "github.com/opencontainers/image-spec/specs-go/v1"
+
+	"github.com/gardener/component-cli/ociclient"
+	"github.com/gardener/component-cli/ociclient/cache"
+	"github.com/gardener/component-cli/ociclient/oci"
+	"github.com/gardener/component-cli/pkg/utils"
+)
+
+const (
+	// ManifestFile is the name of the manifest file of a serialized oci artifact
+	ManifestFile = "manifest.json"
+
+	// IndexFile is the name of the image index file of a serialized oci artifact
+	IndexFile = "index.json"
+
+	// BlobsDir is the name of the blobs directory of a serialized oci artifact
+	BlobsDir = "blobs"
+)
+
+// SerializeOCIArtifact serializes an oci artifact into a TAR archive. the TAR archive contains
+// the manifest.json (if the oci artifact is of type manifest) or index.json (if the oci artifact
+// is a docker image list / oci image index) and a single directory which contains all blobs.
+// the cache instance is used for reading config and layer blobs. returns a reader for the TAR
+// archive which must be closed by the caller.
+func SerializeOCIArtifact(ociArtifact oci.Artifact, cache cache.Cache) (io.ReadCloser, error) {
+	if cache == nil {
+		return nil, errors.New("cache must not be nil")
+	}
+
+	tmpfile, err := ioutil.TempFile("", "")
+	if err != nil {
+		return nil, fmt.Errorf("unable to create tempfile: %w", err)
+	}
+
+	if ociArtifact.IsIndex() {
+		if err := serializeImageIndex(cache, ociArtifact.GetIndex(), tmpfile); err != nil {
+			return nil, fmt.Errorf("unable to serialize image index: %w", err)
+		}
+	} else {
+		tw := tar.NewWriter(tmpfile)
```

Comment:
would expect the NewWriter and tw.Close to be wrapped in the serializeImage func as it is in serializeImageIndex

Reply:
I cannot close the writer in serializeImage because serializeImage is also used for serializing a single image from an image index. If the writer would be closed in serializeImage, I wouldn't be able to use the function for serializing multiple images into the same archive